tests/Security/Authenticator/Oauth2AuthenticatorTest.php

```diff
@@ -0,0 +1,90 @@
+
+use KnpU\OAuth2ClientBundle\Client\OAuth2Client;
+use KnpU\OAuth2ClientBundle\Exception\InvalidStateException;
+use KnpU\OAuth2ClientBundle\Exception\MissingAuthorizationCodeException;
+use KnpU\OAuth2ClientBundle\Security\Authenticator\OAuth2Authenticator;
+use KnpU\OAuth2ClientBundle\Security\Exception\IdentityProviderAuthenticationException;
+use KnpU\OAuth2ClientBundle\Security\Exception\InvalidStateAuthenticationException;
+use KnpU\OAuth2ClientBundle\Security\Exception\NoAuthCodeAuthenticationException;
+use League\OAuth2\Client\Provider\Exception\IdentityProviderException;
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Security\Http\Authenticator\Passport\PassportInterface;
+
+class Oauth2AuthenticatorTest extends TestCase
+{
+    public function testFetchAccessTokenSimplyReturns()
+    {
+        $authenticator = new StubOauth2Authenticator();
+        $client = $this->prophesize('KnpU\OAuth2ClientBundle\Client\OAuth2Client');
+        $client->getAccessToken([])
+            ->willReturn('expected_access_token');
+
+        $actualToken = $authenticator->doFetchAccessToken($client->reveal());
+        $this->assertEquals('expected_access_token', $actualToken);
+    }
+
+    public function testFetchAccessTokenThrowsAuthenticationException()
+    {
+        $this->expectException(NoAuthCodeAuthenticationException::class);
+        $authenticator = new StubOauth2Authenticator();
+        $client = $this->prophesize('KnpU\OAuth2ClientBundle\Client\OAuth2Client');
+        $client->getAccessToken([])
+            ->willThrow(new MissingAuthorizationCodeException());
+
+        $authenticator->doFetchAccessToken($client->reveal());
+    }
+
+    public function testFetchAccessTokenThrowsIdentityProviderException()
+    {
+        $this->expectException(IdentityProviderAuthenticationException::class);
+        $authenticator = new StubOauth2Authenticator();
+        $client = $this->prophesize('KnpU\OAuth2ClientBundle\Client\OAuth2Client');
+        $client->getAccessToken([])
+            ->willThrow(new IdentityProviderException("message", 42, "response"));
+
+        $authenticator->doFetchAccessToken($client->reveal());
+    }
+
+    public function testFetchAccessTokenThrowsInvalidStateException()
+    {
+        $this->expectException(InvalidStateAuthenticationException::class);
+        $authenticator = new StubOauth2Authenticator();
+        $client = $this->prophesize('KnpU\OAuth2ClientBundle\Client\OAuth2Client');
+        $client->getAccessToken([])
+            ->willThrow(new InvalidStateException());
+
+        $authenticator->doFetchAccessToken($client->reveal());
+    }
+}
+
+class StubOauth2Authenticator extends OAuth2Authenticator

```

Comment:
This causes a failure (see the PHP 7.1 tests) if this class doesn't exist in their project (e.g. they are using Symfony 4.4) - https://github.com/knpuniversity/oauth2-client-bundle/pull/292/checks?check_run_id=2168948839

We probably need to wrap this entire `class` declaration in a `if (class_exists(OAuth2Authenticator::class)`. And then we would need to, in `setUp()` for this test class, `markTestSkipped()` if the `OAuth2Authenticator` class doesn't exist.